test/codec-test.hs

```diff
@@ -0,0 +1,271 @@
+  in  testCodecBuilder "SnapshotLabel" $ assembler <$> enumerateSnapshotLabel
+
+testCodecSnapshotTableType :: TestTree
+testCodecSnapshotTableType =
+  let assembler tType =
+        SnapshotMetaData
+          basicSnapshotLabel
+          tType
+          basicTableConfig
+          basicSnapLevels
+  in  testCodecBuilder "SnapshotTableType" $ assembler <$> enumerateSnapshotTableType
+
+testCodecTableConfig :: TestTree
+testCodecTableConfig =
+  let assembler tConfig =
+        SnapshotMetaData
+          basicSnapshotLabel
+          basicSnapshotTableType
+          tConfig
+          basicSnapLevels
+  in  testCodecBuilder "TableConfig" $ assembler <$> enumerateTableConfig
+
+testCodecSnapLevels :: TestTree
+testCodecSnapLevels =
+  let assembler levels =
+        SnapshotMetaData
+          basicSnapshotLabel
+          basicSnapshotTableType
+          basicTableConfig
+          levels
+  in  testCodecBuilder "SnapLevels" $ assembler <$> enumerateSnapLevels
+
+testCodecBuilder :: TestName -> [SnapshotMetaData] -> TestTree
+testCodecBuilder tName metadata =
+    let finalizer :: STRef s Word -> SnapshotMetaData -> ST s TestTree
+        finalizer ref datum = do
+          curr <- readSTRef ref
+          modifySTRef' ref succ
+          pure $ snapshotCodecTest (tName <> printf "-%03d" curr) datum
+
+        testSubtree = runST $ do
+          counter <- newSTRef 0
+          traverse (finalizer counter) metadata
+
+    in  testGroup tName testSubtree
+
+{----------------
+Defaults used when the SnapshotMetaData sub-component is not under test
+----------------}
+
+basicSnapshotLabel :: SnapshotLabel
+basicSnapshotLabel = head enumerateSnapshotLabel
+
+basicSnapshotTableType :: SnapshotTableType
+basicSnapshotTableType = minBound
+
+basicTableConfig :: TableConfig
+basicTableConfig = defaultTableConfig
+
+basicSnapLevels :: SnapLevels RunNumber
+basicSnapLevels = head enumerateSnapLevels
+
+{----------------
+Enumeration of SnapshotMetaData sub-components
+----------------}
+
+enumerateSnapshotLabel :: [SnapshotLabel]
+enumerateSnapshotLabel = [ SnapshotLabel "UserProvidedLabel", SnapshotLabel "" ]
+
+enumerateSnapshotTableType :: [SnapshotTableType]
+enumerateSnapshotTableType = [minBound .. maxBound]
+
+enumerateTableConfig :: [TableConfig]
+enumerateTableConfig =
+    [ TableConfig
+        policy
+        ratio
+        allocs
+        bloom
+        fence
+        cache
+        merge
+    | policy <- [minBound .. maxBound]
+    , ratio <- [minBound .. maxBound]
+    , allocs <- AllocNumEntries . NumEntries <$> [magicNumber1]
+    , bloom <- enumerateBloomFilterAlloc
+    , fence <- [minBound .. maxBound]
+    , cache <- enumerateDiskCachePolicy
+    , merge <- [minBound .. maxBound]
+    ]
+
+enumerateSnapLevels :: [SnapLevels RunNumber]
+enumerateSnapLevels = SnapLevels . V.singleton <$> enumerateSnapLevel
+
+{----------------
+Enumeration of SnapshotMetaData sub-sub-components and so on...
+----------------}
```

Comment:
I would separate the config enumerators from the SnapLevel enumerators in separate sections

Reply:
Moved this under its own header.